### ISSUE TITLE
WPE 2.38 - Video not smooth during initial few seconds for progressive video playback

### ISSUE THREAD

**Issue by pradeep-raveendranpillai-infosys** [Metrological/WebKitForWayland#1309]
On video accelerator devices using wpe 2.38, when progressive video content (e.g. CNN ) is played, it is seen that the video is not smooth (frequent freezes are seen) for the initial few seconds after which video plays smoothly. This issue is not seen with WPE 2.22 version though.

The below workaround is currently undertaken to fix the issue. But it has side effect like video plays in slow speed when the network bandwidth is limited.

--- a/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
@@ -2132,7 +2132,7 @@ void MediaPlayerPrivateGStreamer::updateBufferingStatus(GstBufferingMode mode, d
         m_fillTimer.stop();
 
    +// Force buffering flag to be always false so we don't enter buffering state at all
    +m_isBuffering = false;
    m_bufferingPercentage = percentage;
    switch (mode) {

Sample video is attached.
https://github.com/WebPlatformForEmbedded/WPEWebKit/assets/156415435/373b930e-a06a-44d1-a04b-d83cf52c8360

**Comment by eocanha** [comment]
It's true that buffering in Broadcom devices based on Nexus has been problematic in the past in previous WPE versions, due to the nature of Nexus and how it "sucks" as much data as possible, interfering with the buffering levels measured by WebKit. However, the proposed change is effectively disabling any kind of initial buffering and that doesn't look like a good idea.

I've just tested the change you proposed in WebKit upstream and it breaks these layout tests on my local setup:

http/tests/media/video-play-stall-seek.html
http/tests/media/video-play-waiting.html
media/destructor-logging-crash.html
media/modern-media-controls/button/button-on.html
media/request-video-frame-seek.html

Can you provide a link to the specific video that is causing the problems in your case, as well as a pipeline dump, so I can try to reproduce a similar setup here and analyze how buffering behaves?

In order to get pipeline dumps, you need to define `export GST_DEBUG_DUMP_DOT_DIR=/tmp` and then get the `.dot` files generated on /tmp.

**Comment by pradeep-raveendranpillai-infosys** [comment]
The issue is reproducible using any CNN videos e.g.
https://clips-media-aka.warnermediacdn.com/cnn/clips/2024-04/1501804-2d873fec859e4f478fe6e5b58b29b139/mp4/cms3-CNN-amanpour-brian-greene-primary-356881-1501804-1920x1080_8000k.mp4
https://clips-mp4-aka.warnermediacdn.com/cnn/clips/2024-04/1499525-b1c7ebb4c906452db74112ceee9d8bd6/mp4/cms3-CNN-bandera-texas-eclipse-prep-cowboy-capital-rosa-flores-cncam-primary-356470-1499525-1920x1080_8000k.mp4

Please note that issue might not be reproducible always.

The .dot files and corresponding .png files are attached in a zip file.

[progressive.dot.files.zip](https://github.com/WebPlatformForEmbedded/WPEWebKit/files/15006922/progressive.dot.files.zip)

**Comment by eocanha** [comment]
I haven't been to reproduce the reported behaviour consistently with my (fast) internet connection on a Nexus set-top-box. Still, I've created an experimental [eocanha/eocanha-debug-148](https://github.com/WebPlatformForEmbedded/WPEWebKit/commits/eocanha/eocanha-debug-148/) branch with some commits adapted from https://github.com/WebPlatformForEmbedded/WPEWebKit/issues/792, where a similar problem was reported. The relevant commits there are the most recent three ones.

What that branch does is to have into account the bytes buffered by Nexus' PlayPump, as well as the ones buffered by a multiqueue attached before brcmvidfilter (if any, it wasn't present in the pipelines I got for your test video) and recompute the buffering percentage in a more realistic way.

I got full buffering in 10 sec using a network connection capped to 8000 Mbps (the bandwidth of the test video, 8 Mbit per second = 1 Mbyte per second --> 10 sec to fill the GstQueue2 buffer of 10MB). When the network bandwidth was doubled to 16 Mbps, buffering took 5 sec and I could appreciate more clearly the corrections done by the commits.

If this commits solve your problem, or at least improve them up to a tolerable level, I can work on trying to upstream it.

**Comment by pradeep-raveendranpillai-infosys** [comment]
Thanks, Enrique. 

I tried the last three recent commits from your branch, but it doesn't seem to make any improvement in the behavior. I have attached logs for two video playbacks when the issue happens.

You can please try reproducing the issue using any of the below CNN url. Please try to play as much videos.
https://widgets.metrological.com/lightning/liberty/2e3c4fc22f0d35e3eb7fdb47eb7d4658#app:com.metrological.app.CNN
https://widgets.metrological.com/lightning/rdk/d431ce8577be56e82630650bf701c57d#app:com.metrological.app.CNN

I tried playing same videos on my PC simultaneously to compare the behavior. On PC the issue never happens, while on my STB I can see minimum one pausing occurring after video starts to play.

**Comment by eocanha** [comment]
UPDATE: This request below may not be relevant anymore, since the problem isn't related to how fast the buffering percentage evolves, but related to buffering being considered as complete regardless of the percentage. See the next comment.

I'm going to try, but if you could provide some info about your network bandwidth, or if we could try to simulate the same bandwidth in your environment and in mine, things would be more reproducible.

I'm using Kubuntu 22.04 LTS Jammy Jellyfish as my development environment (but ignore the "K", it's like a regular Ubuntu under the hood) and I'm sharing my internet connection with the set-top-box, so I can run this [bandwidth-limit.txt](https://github.com/WebPlatformForEmbedded/WPEWebKit/files/15039526/bandwidth-limit.txt)
bandwidth shaper script on the second ethernet port that gives access to the STB.

If it's possible for you, run that script on your side and tell me with which bandwidth I should work.

**Comment by eocanha** [comment]
Ok... I see what the problem is in your case (but I still don't know how to solve it, I just know where to look at). The problem is that the pipeline in log1.txt is moving to PLAYING state before the buffering reaches 100%, and that's why underruns happen and the video you posted in the Issue description stutters at the beginning. I'm assuming that video was recorded **without** your patch, right?

The difference with what I can reproduce here is that in my case the "[Buffering] Complete." log line inly happens after the buffering has reached 100%, while in your case it already happens at the begining, even before the percentage is still at 5% (corrected to 16% with playpump).

I'm attaching my (filtered) log for comparison: [filtered-log.txt](https://github.com/WebPlatformForEmbedded/WPEWebKit/files/15039766/filtered-log.txt)

The webpage I use is a simple one that creates a `<video>` and loads it with the raw [cms3-CNN-amanpour-brian-greene-primary-356881-1501804-1920x1080_8000k.mp4](https://clips-media-aka.warnermediacdn.com/cnn/clips/2024-04/1501804-2d873fec859e4f478fe6e5b58b29b139/mp4/cms3-CNN-amanpour-brian-greene-primary-356881-1501804-1920x1080_8000k.mp4) video src you provided.

I still have to check what happens with the CNN app you provided today.

**Comment by pradeep-raveendranpillai-infosys** [comment]
Yes, the video was recorded without my patch (With patch the issue is not there). Even without limiting the network bandwidth I am able to reproduce the issue. My test environment network bandwidth is also very good (> 600 Mbps). But you can try reducing the video bandwidth to 8Mbps - 12 Mbps and see if it helps.

**Comment by eocanha** [comment]
I haven't been able to reproduce the behaviour shown in your logs (and therefore the bug) here, so we'll have to debug this "by email" :-/

Could you please apply the latest https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/9f5ca3d21a644cef33c29def4d4f8e5485cca6d5 commit in the [eocanha/eocanha-debug-148](https://github.com/WebPlatformForEmbedded/WPEWebKit/commits/eocanha/eocanha-debug-148/) branch to your environment and send me new logs? That commits adds some extra logs to try to debug the causes that make m_isBuffering to change. With that information, I can hopefully understand why my environment waits for 100% buffering to start, while yours starts ahead of time, with not enough data buffered, causing the stuttering.

**Comment by pradeep-raveendranpillai-infosys** [comment]
Please find the logs for two of the videos where the issue was reproduced.

**Comment by eocanha** [comment]
I've analyzed your new logs provided by email and I've found that the [query block](https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/9f5ca3d21a644cef33c29def4d4f8e5485cca6d5#diff-4376b854280205494dc9de0f08f985afd4d6584a7487c0a136adac2954c003cdR2631) is being run in your case and not run (at all!) in mine. Also, as I stated in [the commit comments](https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/9f5ca3d21a644cef33c29def4d4f8e5485cca6d5#diff-4376b854280205494dc9de0f08f985afd4d6584a7487c0a136adac2954c003cdR2633), the "buffering percentage == 0 means there's no buffering" condition is playing against us.

I'm going to analyze where that code comes from (what is its purpose) and try to improve it.

**Comment by eocanha** [comment]
I've realized that the behaviour is very different depending if you use playbin2 for regular playback or playbin3 (by setting the `export WEBKIT_GST_USE_PLAYBIN3=1` environment variable). I realized I had been using playbin3 all this time. With playbin2 I reproduced your issue.

I've applied some more corrections that only are into effect for playbin2 (your case). I've tested both variants with successful results here. Could you please try the latest version of the [eocanha-debug-148](https://github.com/WebPlatformForEmbedded/WPEWebKit/commits/eocanha/eocanha-debug-148/) branch?

**Comment by pradeep-raveendranpillai-infosys** [comment]
Thanks, Enrique.

I tested with your changes and the issue seems to be fixed. But, I see the same video stuttering happening when I do a skip/jump ( pressing >> option on the navigation bar of CNN ). 
Also I noticed that the video is taking ~ 5 seconds longer to load initially.

**Comment by eocanha** [comment]
> Also I noticed that the video is taking ~ 5 seconds longer to load initially.

That's because the buffering algorithm was screwed for playbin2 and wasn't waiting at all to have the buffer complete. Now the buffer needs to be actually complete to start playback, and that can take time. You can reduce the percentage on this line to effectively reduce the buffer size and make things a bit more agile, at the expense of bumps when the network conditions are worse.

I've experienced the stuttering you mention after seeks, and sometimes even at the start. It's partially caused by sudden changes between 0% to 100% in the queue2 + playpump + multiqueue corrected joint level, where queue2 was at 0% but playpump was at 100%. I've added a small change to avoid corrections in case queue2 is at 0% (committed to the branch), but some small stuttering still happens because queue2 alternates between 0 and 100% very quickly (0.001s). I still don't have a good explanation about why this happens.

Would the amount of stuttering that is remaining be "acceptabe enough" for you at this point up to the point of considering this a good solution?

Another thing you could try is to use playbin3, by setting the environment variable I mentioned yesterday. Let me know if that may be a better alternative for you.

**Comment by pradeep-raveendranpillai-infosys** [comment]
Even with your latest commit, I am able to see the stuttering happening after a seek and it still doesn't look good.
I even tried with playbin3, there also I could see the stuttering after a seek, though behavior is slightly better when compared to  playbin, but is still not good enough.

**Comment by eocanha** [comment]
Could you please provide logs with the latest changes for the case of seek? Thanks in advance.

**Comment by pradeep-raveendranpillai-infosys** [comment]
Please find the log here

Today while testing, once I saw that after seeking, the video is stuck and never progresses. The log is attached here

**Comment by eocanha** [comment]
I've updated (force-pushed) the [eocanha-debug-148](https://github.com/WebPlatformForEmbedded/WPEWebKit/commits/eocanha/eocanha-debug-148/) branch with changes that...
- Use 80% as the "buffer full" condition
- Use a moving average to smooth the buffer level variability, with some tweaks: The average is reset to 0% on seeks (to ensure rebuffering) and to 100% if a single 100% peak appears (because when queue2 is actually saturated and the pipeline is paused, no more buffering messages will arrive and it won't be possible to get out of the "buffer not full" condition anymore. This will effectively make buffer full to be immediately considered, but buffer draining will take some time (until the average lowers itself).
- Use the buffering level correction in more places.
- **IMPORTANT**: Disallow more that one call to play() and any call to pause() if the environment variable DEBUG_ONLY_ALLOW_FIRST_PLAY=1 is defined. Try these changes first with it defined (you only will be able to play one video on each WPEFramework launch!) and if playback and seek work fine for you, then undefine it. If a lot of suttering or even a permanent stall happens without the variable, then the problem is the webapp JS code issuing play() and pause() calls on its own, which interfere with the buffering mechanism and screw up playback. You should review the webapp in that case. Maybe there's some legacy workaround in the JS app that shouldn't be there.

All these changes mess a lot with the original code. I'm going to start to try to clean them up and even discard those that aren't actually needed (the extra play()/pause() calls by the webapp were really screwing things up during my tests).

Let me know how these changes behave on your side and feel free to provide a new log if they don't work fine. I've observerd a smooth behaviour when network bandwidth isn't limited and buffering times (freezes for initial buffering) of some seconds when it's limited to 16000 Kbps, but that seems reasonable to me.

**Comment by pradeep-raveendranpillai-infosys** [comment]
I tested both with DEBUG_ONLY_ALLOW_FIRST_PLAY and without DEBUG_ONLY_ALLOW_FIRST_PLAY defined. In both cases I could see the issue happening. Attaching the logs.
The issue was more reproducible with the url https://clips-mp4-aka.warnermediacdn.com/cnn/clips/2024-04/1502030-3a5a0e2b43b54391841610ad555354e4/mp4/cms3-CNN-darkness-kerrville-texas-rosa-flores-solar-eclipse-2024-april-8-nasa-vpx-primary-356926-1502030-1920x1080_8000k.mp4

- this is WITHOUT the flag defined. I could see the stuttering after a seek.

- this is WITH the flag defined. I could see the stuttering after a seek, as well as during the initial few seconds as well !!!

The network speed was > 25Mbps when I tested this.

**Comment by eocanha** [comment]
These days I've been working on a hysteresis mechanism for buffering and other improvements, such as unified calculation of m_isBuffering and m_bufferingPercentage, so buffering/not-buffering detection is more consistent. Things work reasonably well here thanks to it (with the patch to ignore subsequent pause()/play() calls by the app) on playbin2. There's still something weird on playbin3 that  I'm currently debugging: 

After seek, sometimes, the player goes "officially" into playing state (spinner disappears), and the GStreamer pipeline follows suit, but there's not enough data ready (buffer isn't full yet), so the pipeline underruns and stalls for some seconds until the data arrives and the picture starts moving. During all that time, WebKit thinks that the video has been playing.

I haven't pushed my local changes to the remote test branch yet.

**Comment by eocanha** [comment]
Although I've been kept working on this full time, I haven't made much progress since the last week report. I have added code to pause the pipeline on seek (so that it doesn't try to blindly continue to play after the seek completes, and instead evaluates the buffering state and only continues when the buffer level is enough), but I still have the issue of the pipeline going to playing state (this includes Nexus gst elements acknowledging that they're in playing state) but Nexus not actually animating the video. There are unexplainable 2.5s of delay (in the worst case) in the audiodecoder and additional 2.5s in the multiqueue that is after it. You can see those delays in the 2nd, 3rd and 4th blocks of these logs generated by my probes (look at the gst log timestamps at the begining of each line):

```
0:00:06.765667931  6131 0x8b105a20 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudfilter0:brcm-audfilter-sink: PTS=0:00:11.660020833, DUR=0:00:00.021333334
0:00:06.767089803  6131 0x8b105a20 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudfilter0:brcm-audfilter-src: PTS=0:00:11.660020833, DUR=0:00:00.021333334
0:00:06.767249385  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudiodecoder0:sink: PTS=0:00:11.660020833, DUR=0:00:00.021333334
0:00:09.283432960  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudiodecoder0:src: PTS=0:00:11.660020833, DUR=0:00:00.021333334
0:00:09.283495845  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### streamsynchronizer0:sink_1: PTS=0:00:11.660020833, DUR=0:00:00.021333334
0:00:09.283530584  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### streamsynchronizer0:src_1: PTS=0:00:11.660020833, DUR=0:00:00.021333334
0:00:09.283838124  6131 0xab9c64c8 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3253:operator(): ### brcmaudiosink0:sink: [DROPPED] PTS=0:00:11.660020833, DUR=0:00:00.021333334

0:00:06.769856292  6131 0x8b105a20 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudfilter0:brcm-audfilter-sink: PTS=0:00:11.681354166, DUR=0:00:00.021333333
0:00:06.771065141  6131 0x8b105a20 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudfilter0:brcm-audfilter-src: PTS=0:00:11.681354166, DUR=0:00:00.021333333
0:00:09.283633505  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudiodecoder0:sink: PTS=0:00:11.681354166, DUR=0:00:00.021333333
0:00:11.793934360  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudiodecoder0:src: PTS=0:00:11.681354166, DUR=0:00:00.021333333
0:00:11.794002801  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### streamsynchronizer0:sink_1: PTS=0:00:11.681354166, DUR=0:00:00.021333333
0:00:11.794042651  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### streamsynchronizer0:src_1: PTS=0:00:11.681354166, DUR=0:00:00.021333333
0:00:11.794119573  6131 0xab9c64c8 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3253:operator(): ### brcmaudiosink0:sink: [DROPPED] PTS=0:00:11.681354166, DUR=0:00:00.021333333

0:00:06.771234389  6131 0x8b105a20 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudfilter0:brcm-audfilter-sink: PTS=0:00:11.702687500, DUR=0:00:00.021333333
0:00:06.772347466  6131 0x8b105a20 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudfilter0:brcm-audfilter-src: PTS=0:00:11.702687500, DUR=0:00:00.021333333
0:00:11.800021633  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudiodecoder0:sink: PTS=0:00:11.702687500, DUR=0:00:00.021333333
0:00:14.312399258  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudiodecoder0:src: PTS=0:00:11.702687500, DUR=0:00:00.021333333
0:00:14.312458848  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### streamsynchronizer0:sink_1: PTS=0:00:11.702687500, DUR=0:00:00.021333333
0:00:14.312491773  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### streamsynchronizer0:src_1: PTS=0:00:11.702687500, DUR=0:00:00.021333333
0:00:14.312805872  6131 0xab9c64c8 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3253:operator(): ### brcmaudiosink0:sink: [DROPPED] PTS=0:00:11.702687500, DUR=0:00:00.021333333

0:00:06.777885297  6131 0x8b105a20 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudfilter0:brcm-audfilter-sink: PTS=0:00:11.724020833, DUR=0:00:00.021333334
0:00:06.778992856  6131 0x8b105a20 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudfilter0:brcm-audfilter-src: PTS=0:00:11.724020833, DUR=0:00:00.021333334
0:00:14.312591843  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudiodecoder0:sink: PTS=0:00:11.724020833, DUR=0:00:00.021333334
0:00:16.825125639  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudiodecoder0:src: PTS=0:00:11.724020833, DUR=0:00:00.021333334
0:00:16.825196970  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### streamsynchronizer0:sink_1: PTS=0:00:11.724020833, DUR=0:00:00.021333334
0:00:16.825230747  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### streamsynchronizer0:src_1: PTS=0:00:11.724020833, DUR=0:00:00.021333334
0:00:16.825329929  6131 0xab9c64c8 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3253:operator(): ### brcmaudiosink0:sink: [DROPPED] PTS=0:00:11.724020833, DUR=0:00:00.021333334

0:00:06.780296994  6131 0x8b105a20 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudfilter0:brcm-audfilter-sink: PTS=0:00:11.745354166, DUR=0:00:00.021333333
0:00:06.781315188  6131 0x8b105a20 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudfilter0:brcm-audfilter-src: PTS=0:00:11.745354166, DUR=0:00:00.021333333
0:00:16.825354150  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudiodecoder0:sink: PTS=0:00:11.745354166, DUR=0:00:00.021333333
0:00:16.886185658  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudiodecoder0:src: PTS=0:00:11.745354166, DUR=0:00:00.021333333
0:00:16.886255915  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### streamsynchronizer0:sink_1: PTS=0:00:11.745354166, DUR=0:00:00.021333333
0:00:16.886292136  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### streamsynchronizer0:src_1: PTS=0:00:11.745354166, DUR=0:00:00.021333333
0:00:16.886391022  6131 0xab9c64c8 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3253:operator(): ### brcmaudiosink0:sink: [DROPPED] PTS=0:00:11.745354166, DUR=0:00:00.021333333

0:00:06.781461068  6131 0x8b105a20 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudfilter0:brcm-audfilter-sink: PTS=0:00:11.766687500, DUR=0:00:00.021333333
0:00:06.782349789  6131 0x8b105a20 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudfilter0:brcm-audfilter-src: PTS=0:00:11.766687500, DUR=0:00:00.021333333
0:00:16.886402133  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudiodecoder0:sink: PTS=0:00:11.766687500, DUR=0:00:00.021333333
0:00:16.886590793  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudiodecoder0:src: PTS=0:00:11.766687500, DUR=0:00:00.021333333
0:00:16.886632866  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### streamsynchronizer0:sink_1: PTS=0:00:11.766687500, DUR=0:00:00.021333333
0:00:16.886662384  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### streamsynchronizer0:src_1: PTS=0:00:11.766687500, DUR=0:00:00.021333333
0:00:16.887176664  6131 0xab9c64c8 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3253:operator(): ### brcmaudiosink0:sink: [DROPPED] PTS=0:00:11.766687500, DUR=0:00:00.021333333

0:00:06.784228817  6131 0x8b105a20 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudfilter0:brcm-audfilter-sink: PTS=0:00:11.788020833, DUR=0:00:00.021333334
0:00:06.785175460  6131 0x8b105a20 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudfilter0:brcm-audfilter-src: PTS=0:00:11.788020833, DUR=0:00:00.021333334
0:00:16.886773084  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudiodecoder0:sink: PTS=0:00:11.788020833, DUR=0:00:00.021333334
0:00:16.886883636  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudiodecoder0:src: PTS=0:00:11.788020833, DUR=0:00:00.021333334
0:00:16.886919635  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### streamsynchronizer0:sink_1: PTS=0:00:11.788020833, DUR=0:00:00.021333334
0:00:16.886945523  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### streamsynchronizer0:src_1: PTS=0:00:11.788020833, DUR=0:00:00.021333334
0:00:16.887241402  6131 0xab9c64c8 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3253:operator(): ### brcmaudiosink0:sink: [DROPPED] PTS=0:00:11.788020833, DUR=0:00:00.021333334

0:00:06.785324895  6131 0x8b105a20 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudfilter0:brcm-audfilter-sink: PTS=0:00:11.809354166, DUR=0:00:00.021333333
0:00:06.786384160  6131 0x8b105a20 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudfilter0:brcm-audfilter-src: PTS=0:00:11.809354166, DUR=0:00:00.021333333
0:00:16.887011150  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudiodecoder0:sink: PTS=0:00:11.809354166, DUR=0:00:00.021333333
0:00:16.887147924  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### brcmaudiodecoder0:src: PTS=0:00:11.809354166, DUR=0:00:00.021333333
0:00:16.887178145  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### streamsynchronizer0:sink_1: PTS=0:00:11.809354166, DUR=0:00:00.021333333
0:00:16.887203848  6131 0xac305928 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3257:operator(): ### streamsynchronizer0:src_1: PTS=0:00:11.809354166, DUR=0:00:00.021333333
0:00:16.887267883  6131 0xab9c64c8 DEBUG      webkitmediaplayer Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:3253:operator(): ### brcmaudiosink0:sink: [DROPPED] PTS=0:00:11.809354166, DUR=0:00:00.021333333
```

The only "explanation" that I've been able to find is that the retry loop in gst_brcm_audio_decoder.c:2709:gst_brcm_audio_decoder_chain() is looping, exhausting its max 82+1 retry iterations and erroring out with the "audio buffer limiting taking longer than expected" warning. I don't know the inner cause for this, because Nexus is sort of a black box for me.

But anyway, even though there's this huge ugly delay after seek, stuttering isn't happening anymore for me. If you wan't I can try to clean all that I have and push it again to the test branch for you to test.

**Comment by eocanha** [comment]
I've updated the [eocanha-debug-148](https://github.com/WebPlatformForEmbedded/WPEWebKit/commits/eocanha/eocanha-debug-148/) branch with my latest changes. It's the most I've been able to get before bumping with the Broadcom Nexus black box, which doesn't give any explanation about why the mysterious 5s delay is happening after seeks.

This branch still has the commit to disable pause() and play() calls. If you enable them, stuttering will appear, but the decision to make those pause() and play() calls comes from the web app and is out of the control of the browser (if the app wants to pause/play multiple times and create stuttering, there's nothing we can do).

If these changes provide improvements good enough, I can start to upstream them and bring them back to the official wpe-2.38 branch.

**Comment by pradeep-raveendranpillai-infosys** [comment]
Thanks, Enrique.
I did limited testing with the changes in your debug branch and below are my observations.

Play and Seek operations
----------------------------
WITHOUT DEBUG_ONLY_ALLOW_FIRST_PLAY defined:
There is no video stuttering during initial playback as well as after a seek. The delay after a seek seems acceptable. Log for the same is -

WITH DEBUG_ONLY_ALLOW_FIRST_PLAY defined:
There is no video stuttering during initial playback. But, after a seek video freezes and never recovers. Log for the same is -


Playback under low network bandwidth condition (2Mbps)
----------------------------------------------------------------
WITHOUT DEBUG_ONLY_ALLOW_FIRST_PLAY defined:
Video doesn't pause for long enough while buffering. Rather it plays slowly i.e. it pauses and resumes too quickly giving a feeling of slow speed playback. Audio drops are heard. This behavior is not good. Log for the same is -

WITH DEBUG_ONLY_ALLOW_FIRST_PLAY defined:
Works fine. Video pauses wile buffering and then resumes. Log for the same is -

**Comment by eocanha** [comment]
Could you please run the "Play and seek operations with DEBUG_ONLY_ALLOW_FIRST_PLAY defined" case with all the debug levels you laready had but, in addition to that, with "webkitmediaplayer:TRACE"? There are some gst messages in that level that I would need in order to understand what's happening and why updateStates() doesn't decide to change the pipeline to playing state again.

**Comment by pradeep-raveendranpillai-infosys** [comment]
vi  /etc/WPEFramework/plugins/WebKitBrowser.json
 "gstdebug":"3,GST_STATES:3,GST_PIPELINE:4,GST_EVENT:4,GST_BUS:4,webkit*:7,brcm*:5,westerosink*:5",

The log for the scenario with trace level set (level 7) is - 
[log.seek.DEBUG_ONLY_ALLOW_FIRST_PLAY.tracelevel.txt](https://github.com/user-attachments/files/15532478/log.seek.DEBUG_ONLY_ALLOW_FIRST_PLAY.tracelevel.txt)

But I could not see any extra logs here :(

**Comment by pradeep-raveendranpillai-infosys** [comment]
Changed debug level to 7 (trace) in /etc/WPEFramework/plugins/WebKitBrowser.json
"gstdebug":"3,GST_STATES:3,GST_PIPELINE:4,GST_EVENT:4,GST_BUS:4,webkit*:7,brcm*:5,westerosink*:5"

The log file for the play scenario with DEBUG_ONLY_ALLOW_FIRST_PLAY defined is given below

**Comment by eocanha** [comment]
An unfortunate mix of events (pipeline forced to pause before seek on purpose, but buffering level evaluating to 100% for some reason after that) made the m_isBuffering true --> false transition to never be detected by updateStates(). That's why playback never resumed.
I've pushed a patch that, on top of forcing the pause of the pipeline before seek, forces re-evaluation of m_isBuffering to true, ensuring (in theory) that the true --> false transition will be detected this time.
Please, run the same test case as before again (and capture logs in case it doesn't work as expected). Let's hope that this time it works...

**Comment by pradeep-raveendranpillai-infosys** [comment]
Tested with DEBUG_ONLY_ALLOW_FIRST_PLAY defined.

Video starts playing without any stuttering, but after a seek it still freezes and never recovers. Log for the same is -


With low n/w conditions (4Mbps) video pauses while buffering and then resumes which is working as expected. But after a seek video freezes and never recovers. Log for the same is - 


For both the logs I have stopped capturing after 10 -15 seconds to limit the log file size, as size > 10MB cant be uploaded here. But the video never recovers for both of the above scenarios, i.e. even after a couple of minutes.

**Comment by eocanha** [comment]
(Sorry, wrong comment, I submitted here a comment that was intended for another issue)

**Comment by eocanha** [comment]
I've updated the eocanha-debug-148 branch with a new change that might prevent the freeze after the seek, by "sinking" the m_isBuffering value to true, so the true to false change is properly detected when buffering is complete. This is something that wasn't happening in the latest test case you reported.

Please retest and let me know any problem that you find this time. Thanks!

**Comment by pradeep-raveendranpillai-infosys** [comment]
DEBUG_ONLY_ALLOW_FIRST_PLAY NOT defined:
Video playback and seek are working under good n/w bandwidth conditions. 
But video playback  under low network bandwidth (2 - 6 Mbps) is not working properly, i.e. video plays slowly with audio drops.

DEBUG_ONLY_ALLOW_FIRST_PLAY Defined:
Video playback and seek are working under good n/w bandwidth conditions. Video playback is working as expected under low network bandwidth conditions (i.e. video pauses while buffering and resumes after) even after seek.

So we need to restrict the webapp from invoking play and pause while buffering?
Also, will the debug code for DEBUG_ONLY_ALLOW_FIRST_PLAY be retained when you push the changes in main branch?

**Comment by eocanha** [comment]
> So we need to restrict the webapp from invoking play and pause while buffering?

Yes, exactly. In any case, knowing why the webapp wants to try to pause/play on its own would be good. Maybe WebKit is generating canplay, canplaythrough, stalled or waiting events in th wrong way (I have not idea, I haven't checked) and whe webapp is reacting on them. Knowing why the webapp reacts like that would be really useful.

> Also, will the debug code for DEBUG_ONLY_ALLOW_FIRST_PLAY be retained when you push the changes in main branch?

No, no, that would break normal usage of the player to watch more than one video. It was just an experimental debug addition to try to isolate the bufering problem without having to worry about the webpage getting on my way.

Talking about pushing changes in the wpe-2.38 branch, I still have to adapt the changes to the new Quirks architecture (a recent addition to avoid so many ifdefs in the code for Broadcom and other kind of devices).

**Comment by pradeep-raveendranpillai-infosys** [comment]
Thanks, Enrique. Since my development branch is not pointing to wpe-2.38 tip (mine is as old as Jan 15th), I shall be importing the commits in your debug branch (except DEBUG_ONLY_ALLOW_FIRST_PLAY related changes) to mine and proceed. Please let me know once you commit the changes to wpe-2.38 branch.

**Comment by eocanha** [comment]
Hmmm... I think in that case you could already import them (except the DEBUG_ONLY_ALLOW_FIRST_PLAY). I say this because the current wpe-2.38 branch has some commits related to the Quirks feature (some refactoring that removes all the Broadcom/Nexus ifdefs and decides about that code at runtime). I've adapted my changes to be compatible with the Quirks framework and I think you would have a hard time de-adapting the new code back.

**Comment by pradeep-raveendranpillai-infosys** [comment]
While doing regression testing with the changes I came across a regression w.r.t. Deezer app - https://widgets.metrological.com/lightning/liberty/2e3c4fc22f0d35e3eb7fdb47eb7d4658#app:com.metrological.app.Deezer

When I select a podcast to play, after it starts playing I am NOT able to Pause the playback. But seek is working.
Attached are the logs - [log.deezer.txt](https://github.com/user-attachments/files/15912582/log.deezer.txt)
The issue was seen only with this app. On other apps like CNN pause is fine.
I suspect the below line as seen from the attached log.
[Buffering] Restarting playback (because of buffering or resuming from zero playback rate)

Another regression was also found on CNN app, i.e. when I give a command in webinspector to set the speed to 2x (document.getElementsByTagName('video')[0].playbackRate = 2), it doesn't play in 2x speed. Log for the same is - 
[log.cnn.2x.speed.txt](https://github.com/user-attachments/files/15912979/log.cnn.2x.speed.txt)
Also, the app is not responding afterwards

**Comment by eocanha** [comment]
The patch has been submitted for upstream review as [bug 275683](https://bugs.webkit.org/show_bug.cgi?id=275683) / https://github.com/WebKit/WebKit/pull/29998.

**Comment by pradeep-raveendranpillai-infosys** [comment]
Okay. Hope the regression issues reported above will also be addressed.

**Comment by pradeep-raveendranpillai-infosys** [comment]
Any update on the fix for the above mentioned regressions?

**Comment by eocanha** [comment]
I've been out of office last week and retook the task yesterday. The Deezer log revealed that the code to manage getting back to play after buffering wasn't having into account voluntary pauses by the user (`m_playbackRatePausedState == PlaybackRatePausedState::ManuallyPaused`). I added a condition to have that into account (locally, not pushed yet) and now need to check if I can reproduce the issue and if that fix works.
I'm also having a look at the CNN log.

**Comment by eocanha** [comment]
I've updated the eocanha-debug-148 branch to include a new commit that should fix the two regressions. Please retest.

Note that the branch has been force-updated. I'm locally working with code adapted to the Quirks framework and have ported the patches to a branch that is still based on ifdefs in the simplest way possible (only using ifdefs when the original code would ask the Quirks framework). I haven't compiled that non-Quirks branch here because I don't want to trigger a full rebuild just for that. It should work.

Now I still have to do all the upstream reworks asked by the reviewers, fix the layout tests there and bring back any fix resulting from that. Dealing with 3 different branches/environments is starting to be messy...

**Comment by pradeep-raveendranpillai-infosys** [comment]
With above changes, I see video stuttering for CNN video after a seek operation. Also there is audio stuttering in Deezer app when playing podcast music (even without seek). The logs for the same is attached. Didn't continue with further testing.
[log.cnn.8jul.txt](https://github.com/user-attachments/files/16123054/log.cnn.8jul.txt)
[log.deezer.8jul.txt](https://github.com/user-attachments/files/16123060/log.deezer.8jul.txt)

**Comment by eocanha** [comment]
Analyzed CNN video stuttering after seek logs. Apparently, the stuttering in the CNN case is legitimately caused by network data starving which brings the buffer below the low watermark level. The Deezer case, on the other hand, is caused by big 0-100% variations in the buffering level. The Broadcom buffer percentage correction isn't applied in this case because Deezer is an audio-only service, and the code looks for vidfilter (not present). I can work on trying to apply the same corrections when only audfilter is present. 

@pradeep-raveendranpillai-infosys, I have a question: Would it be possible for you to upgrade your wpe-2.38 branch to support the Quirks framework (the latest Quirks commits in wpe-2.38)? Maintaining a non-quirks version of the patch is starting to be complex, since [the upstream version of the patch](https://github.com/WebKit/WebKit/pull/29998) is more and more invested in Quirks as the patch evolves, and the more changes there, the more difficul it is to backport the changes to a non-quirks version downstream.

**Comment by eocanha** [comment]
Also... I've just realized that the last version I pushed (without Quirks support) wasn't even compiling because of [this](https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/951227e151f0a39920da3c853f2b915301902f9a#diff-4376b854280205494dc9de0f08f985afd4d6584a7487c0a136adac2954c003cdR529) and [this](https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/951227e151f0a39920da3c853f2b915301902f9a#diff-01c086f2cff46ce9200478ec7952dfcc8a176dbe97369c5c7eff244a6464e40aR720). Sorry about that! (It was because I always worked with the Quirks support version here, and adapted the code to non-Quirks blindly). I've now pushed [a version that builds](https://github.com/WebPlatformForEmbedded/WPEWebKit/commits/668ae39c3af187e230c0b92962fa71e8b0a0c91e).

So... I don't know exactly what version you tried to get those CNN and Deezer problems on the "8jul" logs.

**Comment by pradeep-raveendranpillai-infosys** [comment]
@eocanha I am sorry, upgrading to latest wpe-2.38 may not be possible immediately but eventually it will be done. Sorry to know that you have to do double work here because of that.

Regarding the compilation errors, I had encountered it and I had corrected it myself then. I just forgot to update that in my previous comment.

**Comment by pradeep-raveendranpillai-infosys** [comment]
@eocanha Any update on the fix? 
Does the above commit - https://github.com/eocanha/WebKit/commit/11a45e5d030157ee664fe02ea9f770aad8dbd41f has the necessary changes? If yes, I shall ignore the quirks related changes and take the rest.

**Comment by eocanha** [comment]
I've been out of office the last two weeks, so the only advances have been those in the upstream commit. My two last conclusions regarding the latest problems you reported were:

- The stuttering in CNN was "legitimate" (unstable network?). I couldn't see any strange behaviour of the buffering logic in the logs. No newly added code fixes this.
- The Deezer problem was caused by not having into account audfilter and its PlayPump levels. I've added code (in the upstream PR) to address this.

Migrating the upstream quirks code to your branch isn't so easy, because now all the logic has been moved to the quirks code. My tasks for the next days are:
- Migrating my latest upstream changes downstream (with quirks)
- Migrating the downstream quirks commit to a quirk-less version for you to test
- Debugging upstream layout test failures

**Comment by pradeep-raveendranpillai-infosys** [comment]
@eocanha Any update on the quirk-less version changes?

**Comment by eocanha** [comment]
Not yet. I got busy with more upstream changes asked in the latest review. 🙄

**Comment by eocanha** [comment]
@pradeep-raveendranpillai-infosys I've added a new commit to the [eocanha-debug-148-c-noquirks](https://github.com/WebPlatformForEmbedded/WPEWebKit/commits/eocanha/eocanha-debug-148-c-noquirks/) branch. That should support audio-only pipelines now and hopefully fix the Deezer issue. I can't test it myself because I don't have a test account.

**Comment by pradeep-raveendranpillai-infosys** [comment]
Thanks @eocanha. 
I have started basic testing based the above changes. I found out a minor regression wrt Deezer app podcast (mp3) playback. Seek is now working but the there is a fluctuation in the current time shown just after seek. E.g. When I am at 10th second, if I seek to 40th second, the current time will be shown as 40 first and then it is shown as 11s and then it is shown 40s and is fine thereafter. I am attaching the logs here.

I thought I shall update you whenever I find an issue, rather than waiting for complete testing to be over, so that you can work on it immediately.

By the way this issue is not seen for video playback.

**Comment by eocanha** [comment]
I would need that log, but with `"webkit*:TRACE"` detail level, on top of any other GST_DEBUG string affecting other elements that you might already have. I'm asking for this because [this kind of log lines](https://github.com/WebPlatformForEmbedded/WPEWebKit/blob/wpe-2.38/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp#L1407) (which are TRACE level) are the ones that would help me to understand what's exactly happening with the currentTime/playbackPosition.

**Comment by pradeep-raveendranpillai-infosys** [comment]
Changed debug level to 7 (trace) as given below.
"gstdebug":"3,GST_STATES:3,GST_PIPELINE:4,GST_EVENT:4,GST_BUS:4,webkit*:7,brcm*:5,westerosink*:5"

The log file for the play scenario with DEBUG_ONLY_ALLOW_FIRST_PLAY defined is given below

**Comment by pradeep-raveendranpillai-infosys** [comment]
@eocanha any update on the regression issue?

**Comment by eocanha** [comment]
Sorry, I couldn't work on it during the last two weeks. Will retake it next week.

**Comment by eocanha** [comment]
I've been analyzing the new log you submitted. `playbackPosition()` returns monotonically increasing positions, either [live](https://github.com/WebPlatformForEmbedded/WPEWebKit/blob/wpe-2.38/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp#L1407) or [cached](https://github.com/WebPlatformForEmbedded/WPEWebKit/blob/wpe-2.38/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp#L1401). I can't see any trace of the back-and-forth position change that you describe as a regression. Filtered log lines:

```
0:00:16.518146051 MediaPlayerPrivateGStreamer.cpp:1601:playbackPosition:<video-player-0> Position 0:00:10.945288888, canFallBackToLastFinishedSeekPosition: false
0:00:16.518273606 MediaPlayerPrivateGStreamer.cpp:1596:playbackPosition:<video-player-0> Returning cached position: {10945288888/1000000000 = 10.945288888}
0:00:16.524589051 MediaPlayerPrivateGStreamer.cpp:536:doSeek:<video-player-0> [Buffering] Pausing pipeline, resetting buffering level to 0 and forcing m_isBuffering true before seeking on stream mode
0:00:16.525291051 MediaPlayerPrivateGStreamer.cpp:1596:playbackPosition:<video-player-0> Returning cached position: {10945288888/1000000000 = 10.945288888}
0:00:16.525686828 MediaPlayerPrivateGStreamer.cpp:1596:playbackPosition:<video-player-0> Returning cached position: {10945288888/1000000000 = 10.945288888}
0:00:16.525772125 MediaPlayerPrivateGStreamer.cpp:1596:playbackPosition:<video-player-0> Returning cached position: {10945288888/1000000000 = 10.945288888}
0:00:16.525998865 MediaPlayerPrivateGStreamer.cpp:1596:playbackPosition:<video-player-0> Returning cached position: {10945288888/1000000000 = 10.945288888}
0:00:16.535382791 MediaPlayerPrivateGStreamer.cpp:547:doSeek:<video-player-0> [Seek] Performing actual seek to 0:00:39.769777777 (endTime: 99:99:99.999999999) at rate 1.000000
0:00:16.574406791 WebKitWebSourceGStreamer.cpp:768:webKitWebSrcDoSeek:<source> Seek segment: (537991-18446744073709551615) Position previous to seek: 539735
0:00:20.890542863 MediaPlayerPrivateGStreamer.cpp:2793:updateStates:<video-player-0> [Buffering] Restarting playback (because of buffering or resuming from zero playback rate)
0:00:20.895733381 MediaPlayerPrivateGStreamer.cpp:1596:playbackPosition:<video-player-0> Returning cached position: {39.769777777}
0:00:20.895804270 MediaPlayerPrivateGStreamer.cpp:1596:playbackPosition:<video-player-0> Returning cached position: {39.769777777}
...
0:00:25.245684749 MediaPlayerPrivateGStreamer.cpp:1596:playbackPosition:<video-player-0> Returning cached position: {39795888888/1000000000 = 39.795888888}
0:00:25.245872860 MediaPlayerPrivateGStreamer.cpp:1596:playbackPosition:<video-player-0> Returning cached position: {39795888888/1000000000 = 39.795888888}
0:00:25.246095268 MediaPlayerPrivateGStreamer.cpp:1596:playbackPosition:<video-player-0> Returning cached position: {39795888888/1000000000 = 39.795888888}
0:00:25.451242786 MediaPlayerPrivateGStreamer.cpp:1601:playbackPosition:<video-player-0> Position 0:00:40.004866665, canFallBackToLastFinishedSeekPosition: false
0:00:25.451430564 MediaPlayerPrivateGStreamer.cpp:1596:playbackPosition:<video-player-0> Returning cached position: {40004866665/1000000000 = 40.004866665}
0:00:25.451534082 MediaPlayerPrivateGStreamer.cpp:1596:playbackPosition:<video-player-0> Returning cached position: {40004866665/1000000000 = 40.004866665}
```
```

**Comment by pradeep-raveendranpillai-infosys** [comment]
@eocanha I think the issue may be on the app side. When I tested with Radioline app which is similar to Deezer app, the issue is not seen.
Other issue which I saw is that video playback is slow during low bandwidth network scenario. But if I enable the DEBUG_ONLY_ALLOW_FIRST_PLAY variable, the issue is not seen. I think app is calling play() during buffering in response to "canplay" event which wpe publishes when buffering is complete. You had suggested sometime back to disable app from issuing multiple play(). I wanted to ask you if its logical/possible to add that defensive check in wpe code? i.e. ignoring play/pause while buffering is happening.

Other than the above issues, I couldn't find any regression.

**Comment by eocanha** [comment]
The problem with that defensive check is that it's probably going to break a lot of layout tests. My "debug hack" to disallow play/pause touches WebCore code (multiplatform code) and it can break not only WPE/WebKitGTK, but also Apple ports. I don't think such a defensive check is possible.

If you want, I can try to do the check in a private branch (like the one where I've been placing the commits all this time for you to test) and you can pick it into your downstream repository (at your own risk). Would that work for you?

**Comment by pradeep-raveendranpillai-infosys** [comment]
By the way I was not referring to the "debug hack" which is enabled using DEBUG_ONLY_ALLOW_FIRST_PLAY. I was suggesting something like, in MediaPlayerPrivateGsteamer.cpp::play() we check if m_isBuffering is true i.e. buffering is in progress and ignore the play() request as given below:

    if(m_buffering)
    {
       GST_DEBUG_OBJECT(pipeline(), "Ignoring play from app while buffering");
       return;
    }

I tried this change and I see that the issue is not present. Ignoring play() based on m_isBuffering seems okay to me. I would like to know your comments on this. Can this change also be taken to wpe-2.38 branch?

**Comment by pradeep-raveendranpillai-infosys** [comment]
@eocanha 
1. Can you please comment on my change made on top of your changes for ignoring multiple play requests from app during buffering?

2. Are your other changes (quirks based) for this issue merged to wpe-2.38 main branch?

3. Since LGI branch is pointing to wpe-2.38/Jan 15th date, I had taken the below commits from your dev branch. Please confirm if the below commits is logically equivalent to your quirks based changes.
https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/b8b84bd864bcf74fb114b65ccf7a60691042be64
https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/81fa5789b359b4307d2ba1b90c4557282886c952
https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/627942726c518ef1c5123afe50f3814f227b5333
https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/fa843e7280fd1ecd0cf523de51d9f6f1bc518ecd

**Comment by eocanha** [comment]
1. About your change, I haven't had time to check the implications in all the cases, and I'm a bit scared about unilaterally gnoring play requests causing regressions. For instance, what would happen when the video is prerolling (trying to show its first frame as a preview while still in pause because the JS code hasn't told the player that it wants to play) and buffering because of that, but right in that moment the JS code decides to play()? That play would be ignored and "lost in the void". The JS code might never issue a second play() that would put the video in movement. By now, if possible, I would have more peace of mind if you only added that code to your downstream branch, if possible.

2. No changes are merged yet on any branch because the process is to land the patch upstream first and then backport it. I've been updating a draft version of the backports in the [eocanha-debug-148-c-noquirks](https://github.com/WebPlatformForEmbedded/WPEWebKit/commits/eocanha/eocanha-debug-148-c-noquirks/) and [eocanha-debug-148-c](https://github.com/WebPlatformForEmbedded/WPEWebKit/commits/eocanha/eocanha-debug-148-c/) branches, where I add all the corrections that I do upstream through the review process. My plan is to squash all the commits of those branches and use the original upstream commit message on them when https://github.com/WebKit/WebKit/pull/29998 lands. Unfortunately, and even though I now have r+ from @calvaris as reviewer, some last minute layout test regressions have appeared and I'm working on solving them. From the 3 regressions, I've solved 2 and have only one remaining. I hope to land the patch as soon as possible, but this task has become an oddysey.

3. On a first sight they look equivalent, yes. Did you have any merge conflict when trying to cherry-pick them?

**Comment by pradeep-raveendranpillai-infosys** [comment]
1. I shall see if the change can be limited to my downstream branch.
2. Once you are done will all changes and backport is ready, request you to squash all the non-quirks commits and provide it in a new branch. I shall cherry pick them and after testing, will merge it to LGI's downstream branch.
3. I had been picking and merging the changes manually on my downstream branch till now, so no.

**Comment by pradeep-raveendranpillai-infosys** [comment]
@eocanha can you please give a tentative timeline as to when the changes will be available on wpe-2.38 after downstreaming?

**Comment by eocanha** [comment]
I understand you refer here to "upstreaming" (having the patch landed upstream, which is my primary place of work for the patch and where I'm debugging the only layout test regression remaining). As soon as it lands there, it's just a matter of 2-3 days at most to have it available both on the official 2.38 and on the branch without quirks. I've been porting every partial change I've been doing upstream during the review process, so there's no signifficant downstream effort expected once the patch lands. I wouls just need to squash all the micro-fixes in the downstream branches, remove the debug commits, and reuse the upstream commit message. That's it.

As per when the patch will land upstream, I can't predict it yet, since debugging a layout test is... well, an open task (I don't kow what new challenges I can find). If I had to give an estimation, I would say at the end of next week (W39, September 23-27th 2024), but as any estimation, take it with a grain of salt.

**Comment by eocanha** [comment]
The upstream bug has landed at last as https://commits.webkit.org/284072@main. Porting downstream...

**Comment by pradeep-raveendranpillai-infosys** [comment]
Great!

**Comment by eocanha** [comment]
Ported to wpe-2.38 as https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/de3425aa27719346afe2d58c4c215787b11c170d and also with a variant without quirks as https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/e29d934159d7d1e1b27b01102f8c111498d3fe08 into the https://github.com/WebPlatformForEmbedded/WPEWebKit/tree/eocanha/eocanha-148-noquirks branch (I've removed all the debug commits from it). The [[EXPERIMENTAL] Prevent play() calls while buffering](https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/6a37e6d7c760044e29ab0b72728a8e6c7bd1477a) change has also been added to that branch.

Let me know if we can finally close this Issue.

**Comment by pradeep-raveendranpillai-infosys** [comment]
Thanks.
I shall import the changes in https://github.com/WebPlatformForEmbedded/WPEWebKit/tree/eocanha/eocanha-148-noquirks in my downstream branch. I have below two queries.

1. Do I need to take the commit - https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/13ce063b9a8f0ec52abe8d1b6b2d03fba3f92824
2. The change which I had suggested regarding ignoring multiple play() during buffering, I see that its available in the experimental commit - https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/6a37e6d7c760044e29ab0b72728a8e6c7bd1477a. Wanted to know if it will me merged to upstream & wpe-2.38 branch if no issues found during your testing

**Comment by eocanha** [comment]
You don't need to take commit https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/13ce063b9a8f0ec52abe8d1b6b2d03fba3f92824. That's a commit that I have in the branch I use with the Sagem set-top-box here and I just didn't want to lose it. About the experimental commit, unfortunately no, it won't be integrated into the official wpe-2.38 branch.

**Comment by pradeep-raveendranpillai-infosys** [comment]
@eocanha As discussed before, during low n/w bandwidth, the video playback is slow/stuttering (instead of pausing for sometime and then playing as in the case of good network bandwidth) as app is issuing play() in response to "canplay" event which is sent by wpe after buffering completes. 
As you suggested, one way is to recommend app not to call play() during buffering. But even if app issues play() or not, I feel like its the responsibility of wpe to handle it internally (either by properly ignoring it or not issuing "canplay" while buffering).
Request your inputs on this. If needed, I can create a separate ticket.

**Comment by calvaris** [comment]
Reading from W3C spec:

canplay 	[Event](https://www.w3.org/TR/2011/WD-html5-20110113/infrastructure.html#event) 	The user agent can resume playback of the [media data](https://www.w3.org/TR/2011/WD-html5-20110113/video.html#media-data), but estimates that if playback were to be started now, the [media resource](https://www.w3.org/TR/2011/WD-html5-20110113/video.html#media-resource) could not be rendered at the current playback rate up to its end without having to stop for further buffering of content. 	[readyState](https://www.w3.org/TR/2011/WD-html5-20110113/video.html#dom-media-readystate) newly increased to [HAVE_FUTURE_DATA](https://www.w3.org/TR/2011/WD-html5-20110113/video.html#dom-media-have_future_data) or greater.

If they react to canplay, play, it stops, then buffer, canplay again, play and it stops you get the stuttering but it is perfectly according to the spec. You shouldn't react to can play if you don't have enough buffer. For that you have several other interfaces, like the buffer ranges, the ready states or even:

canplaythrough 	[Event](https://www.w3.org/TR/2011/WD-html5-20110113/infrastructure.html#event) 	The user agent estimates that if playback were to be started now, the [media resource](https://www.w3.org/TR/2011/WD-html5-20110113/video.html#media-resource) could be rendered at the current playback rate all the way to its end without having to stop for further buffering. 	[readyState](https://www.w3.org/TR/2011/WD-html5-20110113/video.html#dom-media-readystate) is newly equal to [HAVE_ENOUGH_DATA](https://www.w3.org/TR/2011/WD-html5-20110113/video.html#dom-media-have_enough_data).

If the original bug is completed as Enrique explained, then I think we can close this.

**Comment by pgorszkowski-igalia** [comment]
> @eocanha As discussed before, during low n/w bandwidth, the video playback is slow/stuttering (instead of pausing for sometime and then playing as in the case of good network bandwidth) as app is issuing play() in response to "canplay" event which is sent by wpe after buffering completes. As you suggested, one way is to recommend app not to call play() during buffering. But even if app issues play() or not, I feel like its the responsibility of wpe to handle it internally (either by properly ignoring it or not issuing "canplay" while buffering). Request your inputs on this. If needed, I can create a separate ticket.

@pradeep-raveendranpillai-infosys : Similar/same issue was raised in the past: https://github.com/WebPlatformForEmbedded/WPEWebKit/issues/831

**Comment by eocanha** [comment]
A follow-up patch to solve some regressions in the debug build (an assert failing) has landed upstream as https://bugs.webkit.org/show_bug.cgi?id=280334 / https://github.com/WebKit/WebKit/pull/34235 / https://commits.webkit.org/284353@main and backported to wpe-2.38 as https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/0dc5f2c09695d2903ce09d7d5ea178c5fc96573d

**Comment by pradeep-raveendranpillai-infosys** [comment]
Thanks @eocanha 

@pgorszkowski-igalia , @calvaris the slow playback issue during bad network is not seen on wpe-2.22 builds. There also app issues multiple play() requests during buffering, in response to "canplay" and "canplaythrough" events sent by wpe (in 2.38 also canplay and canplaythough events are sent). Attaching the logs for wpe-2.22 builds. Can you please comment why it works on 2.22?

**Comment by calvaris** [comment]
There were multiple changes regarding seeking and buffering that could make this not work from one version to the other, but it does not mean that current implementation is not fully compliant with the spec.
